lib/THC/THCTensorMath.cuh

```diff
@@ -23,4 +23,80 @@ __global__ void THCTensor_copyToDiagonal(T* a, T* b, ptrdiff_t start, ptrdiff_t
   }
 }
 
+#define CAT_ARRAY_BATCH_SIZE 1024
+#define CAT_ARRAY_MAX_INPUT_DIMS 4
+
+// Similar to any other IndexToOffset calculation for copying along a given dimension.
+template <typename IndexType, int Dims>
+struct CatArrIndexToOffset {
+  static inline __device__ IndexType compute(
+      const IndexType outputSize[Dims],
+      const IndexType outputStride[Dims],
+      const IndexType dimSize,
+      const unsigned int concatDim,
+      IndexType linearIndex) {
+    IndexType offset = 0;
+
+#pragma unroll
+    for (int i = Dims - 1; i >= 0; --i) {
+      IndexType curDimSize = i == concatDim ? dimSize : outputSize[i];
+      IndexType curDimIndex = linearIndex % curDimSize;
+      IndexType curDimOffset = curDimIndex * outputStride[i];
+      offset += curDimOffset;
+
+      linearIndex /= curDimSize;
```

Comment:
Since both `/` and `%` are being used (and are costly), It may be a little bit more efficient to do something like this to avoid one of them:
```
IndexType nextDimIndex = linearIndex / currDimSize;
IndexType currDimIndex = linearIndex - currDimSize * nextDimIndex;
IndexType currDimOffset = currDimIndex * outputStride[1];
offset += currDimOffset;
linearIndex = nextDimIndex;
```